lib/Dialect/FIRRTL/Transforms/CheckRecursiveInstantiation.cpp

```diff
@@ -0,0 +1,91 @@
+//===- CheckRecursiveInstantiation.cpp - Check recurisve instantiation ----===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/FIRRTL/FIRRTLInstanceGraph.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "mlir/Pass/Pass.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/SCCIterator.h"
+
+namespace circt {
+namespace firrtl {
+#define GEN_PASS_DEF_CHECKRECURSIVEINSTANTIATION
+#include "circt/Dialect/FIRRTL/Passes.h.inc"
+} // namespace firrtl
+} // namespace circt
+
+using namespace circt;
+using namespace firrtl;
+
+static void printPath(InstanceGraph &instanceGraph,
+                      ArrayRef<InstanceGraphNode *> nodes) {
+  assert(nodes.size() > 0 && "an scc should have at least one node");
+
+  // The first node in the SCC where we start the path from.
+  auto *first = nodes.front();
+  auto diag = emitError(first->getModule().getLoc(), "recursive instantiation");
+
+  // A set of all nodes in the SCC for quick lookup.
+  llvm::SmallPtrSet<InstanceGraphNode *, 8> scc(nodes.begin(), nodes.end());
+
+  // A set of all nodes seen in the path.
+  llvm::SmallPtrSet<InstanceGraphNode *, 8> visited;
+
+  // The current node we are examining.
+  auto *current = first;
+  while (true) {
+    // Iterate all instances in this module.
+    for (auto *record : *current) {
+      auto *target = record->getTarget();
+      // Filter out the edges that don't point in the SCC.
+      if (!scc.contains(target))
+        continue;
+      // If this node is already on our path, skip it.
+      if (visited.contains(target))
```

Comment:
```suggestion
      if (!visited.insert(target).second)
```
?

Since this is diagnostic code whatever you think is best.